src/main/java/org/jenkinsci/plugins/vsphere/VSphereConnectionConfig.java

```diff
@@ -140,14 +141,9 @@ public FormValidation doCheckAllowUntrustedCertificate(@QueryParameter boolean v
         public ListBoxModel doFillCredentialsIdItems(@AncestorInPath AbstractFolder<?> containingFolderOrNull,
                 @QueryParameter String vsHost) {
             throwUnlessUserHasPermissionToConfigureCloud(containingFolderOrNull);
-            final Jenkins instance = Jenkins.getInstance(); 
-            if (instance != null) {
-                return new StandardListBoxModel().withEmptySelection().withMatching(
-                    CREDENTIALS_MATCHER, CredentialsProvider.lookupCredentials(StandardCredentials.class,
-                    instance, ACL.SYSTEM, getDomainRequirement(vsHost))
-                );
-            }
-            return new StandardListBoxModel();
+            return new StandardListBoxModel().includeEmptyValue()
+                .includeMatching(Jenkins.getInstance(), StandardCredentials.class,
+                    Collections.singletonList(getDomainRequirement(vsHost)), CREDENTIALS_MATCHER);
```

Comment:
With this change, I can no longer select any credentials since the dropdown list only contains “-none-“. I reverted the change to this method and the dropdown list is properly populated.
I would like to understand why the change was made and if the problem I am seeing is a bug or not. Any feedback you can supply would be appreciated.

Reply:
Hmm.... my _guess_ is that you don't have sufficient rights on your Jenkins server to see all the credentials you're used to seeing.
The old code specified
```
                return new StandardListBoxModel().withEmptySelection().withMatching(
                    CREDENTIALS_MATCHER, CredentialsProvider.lookupCredentials(StandardCredentials.class,
                    instance, ACL.SYSTEM, getDomainRequirement(vsHost))
```
whereas the new code is missing the `ACL.SYSTEM` bit, so I would _guess_ that it's showing you what credentials you're allowed to see instead of all the credentials the system can see.
...but that's a guess - I'm not hugely familiar with how `StandardListBoxModel` is (or should be) used so I can't tell the difference between "good enough that it'll work in my own local environment" and "actually correct".
I can tell you that it "works for me" - if I go to "manage jenkins" -> "configure system", scroll down to "clouds" and then click on the drop-down list of credentials for a specific vSphere cloud then I see lots of credentials ... _but_ I'm logged as "as administrator" and I don't have any restrictions on how my credentials are used.  My assumption is that your setup differs, which is why it's not working for you.

If we can get to the bottom of this, I'd be quite happy to receive a PR that fixes it for you (as long as it doesn't break it for anyone else)...

Reply:
I believe for this to work with folder credentials it might need some tweaking

Reply:
Ah, yes, folders make everything vastly more complicated, and they're not something I have any experience with.

If you are able to try out different options ... try using the new code but where it says `instance` put instead `containingFolderOrNull!=null?containingFolderOrNull:instance`
That should then provide the folder as the "context" for the credentials list instead of the Jenkins administrator-only list.

...and if that works then I suspect that we'll need similar changes elsewhere too...

Reply:
This change seems to be the cause of [JENKINS-61131](https://issues.jenkins-ci.org/browse/JENKINS-61131).
Can please a maintainer take a look at my proposed fix in #116?

Reply:
Looked; merged.

@albers Given that you know enough on this subject to be able to code a fix ... what's your opinion on `containingFolderOrNull!=null?containingFolderOrNull:Jenkins.getInstance()` vs `Jenkins.getInstance()` ?
My own experience with credential lists is more "shooting in the dark" than actual informed opinion so if you know enough to suggest fixes you probably know more about how this _should_ be done than I do - I'd welcome any informed suggestions, especially if it'll let us get it right this time - trial and error isn't an approach I like to use...

Reply:
ContainingFolderOrNull is not the proper way of doing it.

From jenkins you can ask for the context using `@AncestorInPath Item item`

https://github.com/jenkinsci/git-plugin/blob/c0e8144da63e3c8ccea7f00f7438b6e70aac9b00/src/main/java/hudson/plugins/git/UserRemoteConfig.java#L99-L121

Reply:
While I am familiar (as in, I've seen it used before) with `@AncestorInPath Item item`, I'm less certain about `Tasks.getAuthenticationOf((Queue.Task) project)`; I suspect that the vSphere plugin is unlikely to encounter a `Queue.Task` and that this is more specific to the git plugin's use-case...

The existing vsphere-plugin code already obtains `@AncestorInPath AbstractFolder<?> containingFolderOrNull` and as `AbstractFolder` is `instanceof Item` I would expect `containingFolderOrNull` to be the same value we'd get if this code had requested `@AncestorInPath Item item` instead.
What I'm less certain of is exactly what to do with it...